include/sparrow/buffer.hpp

```diff
@@ -18,454 +18,870 @@
+    {
+        const size_type sz = size();
+        const size_type len = std::distance(first, last);
+        if (len > capacity())
+        {
+            check_init_length(len, get_allocator());
+            pointer p = allocate_and_copy(len, first, last);
+            destroy(get_data().p_begin, get_data().p_end, get_allocator());
+            this->deallocate(get_data().p_begin, capacity());
+            this->assign_storage(p, len, len);
+        }
+        else if (sz >= len)
+        {
+            pointer p = std::copy(first, last, get_data().p_begin);
+            erase_at_end(p);
+        }
+        else
+        {
+            It mid = first;
+            std::advance(mid, sz);
+            std::copy(first, mid, get_data().p_begin);
+            get_data().p_end = copy_initialize(mid, last, get_data().p_end, get_allocator());
+        }
     }
 
     template <class T>
-    buffer<T>& buffer<T>::operator=(buffer<T>&& rhs)
+    constexpr void buffer<T>::erase_at_end(pointer p)
     {
-        base_type::swap(rhs);
-        return *this;
+        destroy(p, get_data().p_end, get_allocator());
+        get_data().p_end = p;
     }
 
     template <class T>
-    void buffer<T>::resize(size_type n)
+    template <class It>
+    constexpr auto buffer<T>::allocate_and_copy(size_type n, It first, It last) -> pointer
     {
-        // TODO: add capacity, resize if growing only and define a shrink_to_fit method
-        if (n != base_type::size())
+        pointer p = this->allocate(n);
+        try
         {
-            buffer<T> tmp(n);
-            const size_type copy_size = std::min(base_type::size(), n);
-            std::copy(base_type::data(), base_type::data() + copy_size, tmp.data());
-            base_type::swap(tmp);
+            copy_initialize(first, last, p, get_allocator());
         }
+        catch(...)
+        {
+            this->deallocate(p, n);
+            throw;
+        }
+        return p;
     }
 
     template <class T>
-    void buffer<T>::resize(size_type n, value_type value)
+    constexpr auto buffer<T>::check_init_length(size_type n, const allocator_type& a) -> size_type
     {
-        const size_type old_size = base_type::size();
-        resize(n);
-        if (old_size < n)
+        if (n > max_size_impl(a))
         {
-            std::fill(base_type::data() + old_size, base_type::data() + n, value);
+            throw std::length_error("cannot create buffer larger than max_size()");
         }
+        return n;
     }
 
     template <class T>
-    void buffer<T>::clear()
+    constexpr auto buffer<T>::max_size_impl(const allocator_type& a) noexcept -> size_type
     {
-        resize(size_type(0));
+        const size_type diff_max = std::numeric_limits<difference_type>::max();
+        const size_type alloc_max = std::allocator_traits<allocator_type>::max_size(a);
+        return (std::min)(diff_max, alloc_max);
```

Comment:
Why is this needed?
```suggestion
        return std::min(diff_max, alloc_max);
```

Reply:
Because of the min macro defined on Windows, but we can remove it if we define `NOMINMAX` and propagate it.